### ISSUE TITLE
ES6 style - `import tr from 'transduce';` results in undefined?? while var tr = require('transduce') works

### ISSUE THREAD

**Issue by jeffbski** [transduce/transduce#56]
Surprising issue, if I use ES6 style import with babel (compiled to commonjs) to import transduce it comes back as undefined.

```javascript
import tr from 'transduce';
console.log(tr); // undefined
```

while using normal require works fine

```javascript
var tr = require('transduce');
console.log(tr); // object as expected
```

I have been using ES6 style imports with babel (that are compiling down to commonjs) for most everything and this is the only project I have had this trouble with. 

Is there something about how things are exported that makes this incompatible?

Is there anything we can do to fix it so that either way will work?

PS. I love the project.

Thanks,

Jeff

**Comment by kevinbeaty** [comment]
Hmm... yeah that is surprising.  ES6 was recently added to this project and it's my first experience with Babel, so I'm no really sure why that is failing.  @tgriesser Any ideas?

> PS. I love the project.

Thanks Jeff!

**Comment by kevinbeaty** [comment]
I think this is related to how Babel handles `interopRequire` with default exports.  There is no default export in `transduce/index` but it is marked as a `__es6module`, so 

```javascript
import tr from 'transduce';
```

Returns the default export with `interopRequire`.

A wildcard import would work

```javascript
import * as tr from 'transduce'
```

Or (my preference), import using destructuring:

```javascript
import {map, filter} from transduce
```

**Comment by jeffbski** [comment]
ok, that explains it. I wonder if we could have a default export so that it wouldn't be an issue regardless. 

I can imagine people just trying to follow the examples and being surprised when it doesn't work (and possibly abandoning the project since they don't know how to resolve quickly). Yes, I know the examples use require, but as someone who is trying to do more things the ES6 way, I just substitute import for the require and normally it works. They would have to know the intricacies of how to debug etc.

So if it is possible to have a default, then if they naively do the import as I did, it would work regardless.

**Comment by kevinbeaty** [comment]
> I can imagine people just trying to follow the examples and being surprised when it doesn't work

I agree, I was too...

>  but as someone who is trying to do more things the ES6 way, I just substitute import for the require and normally it works

I am one of those people :)  ... Like I said, this is the first time I've tried ES6.  Is it common for the default export to be set to the module?  Is this the "ES6 way" to import modules or is this an accident of the babel interop integration with CommonJS?  I'm not opposed to setting it if it's common practice (although I'm not sure how it's done).  There is no obvious choice for a default export except the module itself in this case.

**Comment by jeffbski** [comment]
My thoughts...

With the transducer library, it would be my assumption that the default export would be the top level module, just like it is when using commonjs require. Personally, I can't think of anything else that would be more appropriate than this. 

**So I think we would want the default import to do something useful in any case.** That is the **sweetest syntax to use and so we wouldn't want to ignore it and return undefined**. If there was something more appropriate to return than the top module we could, but given the commonjs roots of this project, it would seem that it is the likely choice. If there was a better choice for the default, then we would want to heavily document this so others are not surprised.

More background which can help with this decision is available below:

The best info I am aware on ES6 modules is here: http://www.2ality.com/2014/09/es6-modules-final.html 

> "ECMAScript 6 favors the single/default export style, and gives the sweetest syntax to importing the default. Importing named exports can and even should be slightly less concise."

**Comment by kevinbeaty** [comment]
>  it would be my assumption that the default export would be the top level module

I agree, if we were exporting a default, this would be the most appropriate.

> That is the sweetest syntax to use 

I actually think the destructuring supported by named exports/imports is the "sweetest syntax", so that's probably mostly preference

> So I think we would want the default import to do something useful in any case.
> ... so we wouldn't want to ignore it and return undefined.

If we could support both a default export of the whole module and the named export/import, that would be fine by me.  Is there a way in ES6 to say "export an object with all named exports as a default"? (I don't think there is, but would be happy to be wrong)  Or maybe a way in Babel to support both for interop?  Or maybe another option? 

In sum, I'd be OK with adding the support for exporting the module as "default" but I don't want to give up the named imports.

In any case, it wouldn't hurt to add some ES6 examples for documentation.

**Comment by jeffbski** [comment]
Would something like this work?

```javascript
import * as core from './core'
import * as transducers from  './transducers'
import * as array from  './array'
import * as string from  './string'
import * as math from  './math'
import * as async from  './async'
const modules = {array, string, math, async};
const transduce = Object.assign({}, core, transducers, modules);
export default transduce;
```

I haven't tested this, but I believe this accomplishes what we have been discussing.

It provides a default object with core and transducers properties directly on the object, and the array, string, math, async are one level down. 

You can also still use the named imports like `import { into, async } from 'transduce';`

**Comment by tgriesser** [comment]
Yeah, exporting in this way is a notably missing feature in ES6 - proposed for inclusion in ES7 here: https://github.com/leebyron/ecmascript-more-export-from

I think what @jeffbski is proposing should work though

**Comment by mhuggins** [comment]
I'm having a similar issue with a custom set of files.  I **am** defining my export as a default export, but it makes no difference.

```javascript
# BaseClient.js
export default class BaseClient {
  constructor() {
    this.rootURL = ROOT_URL;
  }

  // snip...
};

// SessionClient.js
import BaseClient from './BaseClient';

console.log('BaseClient', BaseClient);  // outputs 'undefined'

export default class SessionClient extends BaseClient {
  // ^^^ throws "Uncaught TypeError: Super expression must either be null or a function, not undefined"
};
```
